MRG: Refactor time frequency transform

kingjr opened the pull request:
- [X] Moved functions around inside `mne/time_frequency/tfr.py` to clarify what refers to 1) wavelet definition 2) computation of convolution 3) numpy array wrappers 4) single trial versus averages 5) instance related (classes, picks etc). (This is the first commit, nothing is modified, just moved).
- [X] Added a single function `timefreq_transform` to compute both single trial and 'averages' time frequency transforms
- [X] Added parallelization across channels (Makes more sense than epochs because can be applied to both single trial and average metrics, as well as to Epochs-type of data and Raw-types of data)
- [X] Checked that the results are numerically identical to the previous functions, concatenate tests to new function.
- [X] Deprecated `cwt_morlet`, and `single_trial_power`, remove `_induce_*` functions.
- [X] Change `_dpss_wavelet` into `_make_dpss` and `morlet` into `_make_morlet`

What won't be addressed in this PR:
- deprecate `tfr_morlet` and `tfr_mtm` into a single function
- clean up instance related function and classes
- add wrapper to apply to raw data, or 2D numpy arrays. (For 2D numpy arrays, I think it is safer to ask the user to ensure a 3D array where the first dimension is of length 1, rather than doing fancy conditional input).
- add array_split parallelization to reduce overhead (easy, but needs to be benchmarked)
- I currently added a `NotImplementedError` error for single trial phase estimation with mtm. I'm not sure whether this computation is actually valid.
- deprecation of `mne.time_frequency.cwt()` which is similar to the new function except that it allow passing `Ws` directly. This is used by `mne.connectivity.spectral`, and could be deprecated in a subsequent PR.

What remains to be addressed
- [X] compatibility with mne objects calling the now deprecated function.
- [ ] benchmark parallel (I'm not very good at that though...)

closes https://github.com/mne-tools/mne-python/pull/3445

Since it's a lot of lines, I suggests to do the review per commit, I tried to keep them clean.

mmagnuski commented on mne/time_frequency/tfr.py:
```diff
@@ -227,6 +227,237 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    frequencies = np.asarray(frequencies)
+
+    # XXX Check not sure whether it makes sense to compute single-trial phases
+    # with mtm.
```
I remember reading fieldtrip guys saying it's better not to do it because of the frequency smoothing.

kingjr replied:
Yes, this is what I realize while implementing it. But I'm not sure whether we should just throw a warning or actually prevent users doing it.

It wasn't possible to compute single trial phase with mtm in the previous code, so I guess we can leave this for another PR

mmagnuski replied:
maybe `BadIdeaError`? :)

mmagnuski commented on mne/time_frequency/tfr.py:
```diff
@@ -227,6 +227,237 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+
+    # Check wavelets
+    if len(Ws[0][0]) > epoch_data.shape[2]:
+        raise ValueError('Wavelet is too long for such a short signal.')
```
I was recently thinking it might be better phrased as "At least one of the wavelets is longer than the signal. Use longer signal or shorter wavelets."

mmagnuski commented on mne/time_frequency/tfr.py:
```diff
@@ -227,6 +227,237 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    # Parallel computation
+    parallel, my_cwt, _ = parallel_func(_time_frequency_loop, n_jobs)
+
+    # Loop across different wavelet in case of multitaping
```
"across different **channels**" now

mmagnuski replied:
ah, no, sorry, didn't check parallel_func

mmagnuski commented:
I looked through the code in a hurry (I'll take a close look later) but I like the changes - the time-frequency code was a little messy before.

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    # Parallel computation
+    parallel, my_cwt, _ = parallel_func(_time_frequency_loop, n_jobs)
+
+    # Loop across different wavelets in case of multitaping
```
needs fix

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        raise ValueError('sfreq must be a float or an int.')
+    sfreq = float(sfreq)
+
+    # Check params
```
needs fix

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+
+
+def _time_frequency_loop(X, Ws, output, use_fft, mode, decim):
+    """Aux. function to time_frequency.
```
needs fix

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+
+    Parameters
+    ----------
+    X: np.array, shape (n_epochs, n_times)
```
needs fix

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -486,75 +707,131 @@ def _induced_power_cwt(data, sfreq, frequencies, use_fft=True, n_cycles=7,
+        Return intertrial coherence (ITC) as well as averaged power.
+        Defaults to True.
+    decim : int | slice
+        To reduce memory usage, decimation factor after time-frequency
```
time frequency

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -486,75 +707,131 @@ def _induced_power_cwt(data, sfreq, frequencies, use_fft=True, n_cycles=7,
+        If `int`, returns tfr[..., ::decim].
+        If `slice` returns tfr[..., decim].
+        Note that decimation may create aliasing artifacts.
+        Defaults to 1.
```
1

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -486,75 +707,131 @@ def _induced_power_cwt(data, sfreq, frequencies, use_fft=True, n_cycles=7,
+        Note that decimation may create aliasing artifacts.
+        Defaults to 1.
+    n_jobs : int
+        The number of jobs to run in parallel. Defaults to 1.
```
1

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -486,75 +707,131 @@ def _induced_power_cwt(data, sfreq, frequencies, use_fft=True, n_cycles=7,
+    power : AverageTFR
+        The averaged power.
+    itc : AverageTFR
+        The intertrial coherence (ITC). Only returned if return_itc
```
inconsistent naming: phase locking factor, intetrial coherence

agramfort replied:
I first called this phase lock but Robert from fieldtrip convinced me that intertrial coherence is the proper name. We should use the latter

mmagnuski replied:
You can also consider ITPC (inter-trial phase coherence vel. inter-trial phase clustering) ;)
I prefer ITC btw, it is used frequently (mostly in eeglab)

kingjr replied:
ok, shall I change the option name into `avg_power_itc` ? What do you guys suggest? It's hard to find a systematic nomenclature..

mmagnuski replied:
`itc` and `avg_power_itc` should be ok (`avg_power&itc`? or maybe if you wan't both avg_power and itc it should be `['avg_power', 'itc']`?)

dengemann replied:
@mmagnuski we had called it already `power` for a long time, we should not change names just so. I think that with the docs it should be clear. We changed some time ago `plf` to `itc`, we should see if some remainder is still existing.

kingjr commented on mne/time_frequency/tfr.py:
```diff
@@ -1130,12 +1407,157 @@ def save(self, fname, overwrite=False):
-                            nave=tfr.nave, comment=tfr.comment,
-                            method=tfr.method))
+def combine_tfr(all_tfr, weights='nave'):
+    """Merge AverageTFR data by weighted addition
```
addition.

kingjr commented:
I don't understand the travis error and don't manage to replicate locally. Anyone to help?

mmagnuski commented:
Seems to be just docstring formatting error:
```
mne.utils.single_trial_power arg mismatch: ['data']
>>  raise AssertionError("\nmne.utils.single_trial_power arg mismatch: ['data']")
```

kingjr commented:
@mmagnuski yes I have seen, but what exactly is problematic

mmagnuski commented:
Ah, right, weird...

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    ----------
+    X : array, shape (n_epochs, n_times)
+        The epochs data of a single channel.
+    Ws : list, shape(n_tappers, n_wavelets, n_times)
```
space after shape

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -459,102 +677,156 @@ def single_trial_power(data, sfreq, frequencies, use_fft=True, n_cycles=7,
+def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
+                   use_fft=True, return_itc=True, decim=1,
+                   n_jobs=1, picks=None, verbose=None):
+    """Compute Time-Frequency Representation (TFR) using DPSS wavelets
```
DPSS wavelets -> wavelets and DPSS tappers

agramfort commented on doc/whats_new.rst:
```diff
@@ -43,6 +43,8 @@ Changelog
 
     - Add the ``--no-decimate`` option to :ref:`mne make_scalp_surfaces <gen_mne_make_scalp_surfaces>` to skip the high-resolution surface decimation step, by `Eric Larson`_
 
+    - Added :func:`mne.time_frequency.time_frequency` and deprecated :func:`mne.time_frequency.cwt_morlet` and :func:`mne.time_frequency.single_trial_power`, to compute time frequency transforms on arrays, by `Jean-Remi King`_
```
time_frequency -> tfr_transform

agramfort commented on mne/time_frequency/tests/test_tfr.py:
```diff
@@ -212,8 +212,8 @@ def test_time_frequency():
     freqs = np.arange(5, 25, 3)
-    Ws = _dpss_wavelet(1000, freqs=freqs, n_cycles=freqs / 2.,
-                       time_bandwidth=4.0, zero_mean=True)
+    Ws = dpss_wavelet(1000, freqs=freqs, n_cycles=freqs / 2.,
```
why making dpss_wavelet public?

mmagnuski replied:
`morlet` was public and dpss_wavelet not so either both should be private or both public.

kingjr replied:
I'll make both of them private, less support-related work for for the future

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+def tfr_transform(epoch_data, frequencies, sfreq=1.0, method='morlet',
+                  n_cycles=7.0, zero_mean=None, time_bandwidth=4.0,
+                  use_fft=True, decim=1, output='complex', n_jobs=1,
+                  verbose='INFO'):
```
needs @ verbose decorator and don't force verbose to 'INFO'

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        Sampling frequency of the data.
+    method : 'mtm' | 'morlet', defaults to 'morlet'
+        The time frequency method. 'morlet' convolves a Morlet wavelet. 'mtm'
+        convolves DPSS wavelets using a multitaper approach.
```
DPSS are the tapers that are multlplied elementwise to the data. DPSS are not wavelets.

mmagnuski commented on mne/time_frequency/tfr.py:
```diff
@@ -29,52 +30,39 @@
-def morlet(sfreq, freqs, n_cycles=7, sigma=None, zero_mean=False):
-    """Compute Wavelets for the given frequency range
+def morlet(sfreq, freqs, n_cycles=7.0, sigma=None, zero_mean=False):
+    """Compute wavelets for the given frequency range.
```
"Compute **Morlet** wavelets ..."

agramfort commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        my_cwt(channel, Ws, output, use_fft, 'same', decim)
+        for channel in epoch_data.transpose(1, 0, 2))
+
+    # FIXME: to avoid overheads we should use np.array_split()
```
yeah something like this. I suspect this code is pretty greedy in terms of memory usage

kingjr replied:
Why do you say that? The data is split across cores with no redundancy, so there is no RAM overload. Also, because we're looping across channels, we can apply the average metrics without having to store single trial tfr transforms.

I'm adding this FIXME so that we can, in another PR, check whether an array splitting would be worth implementing. It's an additional loop and stacking; not really difficult, but I'd like to minimize my work here.

Eric89GXL replied:
> The data is split across cores with no redundancy, so there is no RAM overload.

When using joblib parallelization, the data are *pickled* and sent to each core. So in theory, you have now made an entire copy of your data, and sent it out.

When using a single core / no parallelization, you can get away with just sending *views* of the data usually, and often only having a copy of a single channel at a time.

Eric89GXL replied:
...and so each time you create a new job, all data must be pickled and sent out. So if you have two arguments AA and B, and AA comes from iterating over A, if B is also large there is an advantage to sending A split into only `n_jobs` chunks (jobs) as opposed to A split into 306 chunks (jobs), because the former only needs to pickle B `n_jobs` times, and the latter does it 306 times.

kingjr replied:
> When using a single core / no parallelization, you can get away with just sending views of the data usually, and often only having a copy of a single channel at a time.

If there's a single job, I thought mne.parallel was basically skipping joblib; does it not?

> So if you have two arguments AA and B, and AA comes from iterating over A, if B is also large there is an advantage to sending A split into only n_jobs chunks (jobs) as opposed to A split into 306 chunks (jobs), because the former only needs to pickle B n_jobs times, and the latter does it 306 times.

IIUC, you're saying that splitting the data into n_chunks, where each chunk has several channels, then this will be more memory consuming than splitting the data into 306 jobs directly. If so, yes I agree. It is a memory / CPU trade-off.

In any case, I think splitting over channels is a better strategy than over epochs as it was before, because it allows computing single trial or inter-trial metrics.

Eric89GXL replied:
> If there's a single job, I thought mne.parallel was basically skipping joblib; does it not?

Yes, but depending on the array sizes it can end up being more efficient to triage based on `n_jobs==1` or not. Here I doubt it will matter much in terms of memory as it is currently coded.

> IIUC, you're saying that splitting the data into n_chunks, where each chunk has several channels, then this will be more memory consuming than splitting the data into 306 jobs directly. If so, yes I agree. It is a memory / CPU trade-off.

Yeah, there is a tradeoff. Have you looked at the memory and CPU use for multiple use cases for `n_jobs=1` vs `n_jobs=-1`, both when using direct iteration as here and when using `np.array_split` of the data? Sometimes the pickling overhead is really bad, so the additional memory usage is worth it. It would take some testing to know.

> In any case, I think splitting over channels is a better strategy than over epochs as it was before, because it allows computing single trial or inter-trial metrics.

This could be true, it could also be slower again because of the pickling. Now if there are 100 epochs and 306 channels, you're pickling certain parameters 30,000+ times instead of only `n_jobs` times, which could be brutal. Again, it's worth checking to see how it does. We might end up needing two paths of parallelization, one for multiple trials (or more than `N` trials) and one for single-ish trials.

TL;DR: I see how this could be beneficial in some situations, but without someone doing the benchmarks, it's hard to know how often this is beneficial vs detrimental w/r/t CPU and memory tradeoffs.

kingjr replied:
I agree. If someone hits optimization issues maybe he/she can benchmark the two routes options.

> Now if there are 100 epochs and 306 channels, you're pickling certain parameters 30,000+ times instead of only n_jobs time

This I didn't get. In the worse case scenario, we pickle 306 times don't we?

Eric89GXL replied:
> I agree. If someone hits optimization issues maybe he/she can benchmark the two routes options.

But it used to iterate the other way, right? In other words, your PR changes the way the parallelization is done, and therefore could introduce speed regressions for some use cases, and that hasn't been checked...?

>  In the worse case scenario, we pickle 306 times don't we?

Ahh I thought you were iterating over `n_epochs * n_channels` 1D (time) arrays, but I see now you're iterating over `n_channels` 2D (epoch x time) arrays. So yeah 306. In that case, I suppose the iteration order shouldn't matter too much, assuming there isn't a huge difference in the number of channels vs the number of epochs people will tend to have. Is that what you have in mind?

kingjr replied:
> But it used to iterate the other way, right? In other words, your PR changes the way the parallelization is done, and therefore could introduce speed regressions for some use cases, and that hasn't been checked...?

I haven't done the benchmarking, because I'm running out of time with the GSoC-related PR, but it should be  quite similar to what we had before. The parallelization was either absent or across channels too for inter trials metrics. single_trial_power parallelized across epochs so I guess this one could change the computation.

> Is that what you have in mind?

yes

agramfort commented on tutorials/plot_stats_cluster_1samp_test_time_frequency.py:
```diff
@@ -72,17 +73,20 @@
-                                  n_cycles=4, n_jobs=1,
-                                  baseline=(-100, 0), times=times,
-                                  baseline_mode='ratio', decim=decim)
+epochs_power = tfr_transform(data, frequencies, sfreq=sfreq, n_cycles=4.,
```
these stats examples are the reason why tfr_transform is public AFAIK

would it possible make use of tfr_morlet directly?

I would be more encline to merge if we manage to have tfr_transform private

mmagnuski replied:
I think it is nice to have one public interface for time-frequency, similar to `ft_freqanalysis` in fieldtrip. Also - tfr_transform seems to me to be too high-level to be private.

kingjr replied:
I can make `tfr_transform` private if you really want, but I'll let you take care of the refactoring of high level objects that detect the input type, pick the right channels etc. The main annoyance for me is that the function you're thinking of returns an MNE object (currently, `AverageTFR`, or tuple of `AverageTFR`, and we'd need to create a new class for single trials etc, deal with 'complex' versus 'power' or  'inter-trial coherence' output types etc).

I'm personally against these these kind of heavy object approaches. The people I work with generally complain that most package enforce using complex structures to be able to use their functions, and therefore just re-code them directly. IMO, time frequency transform should be usable with numpy arrays: we don't need to know whether it's an MEG or an EEG channels, it's layout etc, many users just want a computationally efficient convolution.

mmagnuski commented:
Just two things about naming:
* if we have `dpss_wavelet` maybe have `morlet_wavelet` instead of just `morlet`?
* `tfr_transform` sounds a little weird - at least if `tfr` stands for time-frequency *representation*. time-frequency representation transform? maybe `tf_transform`, `tfr` or `time_frequency`? I think `time_frequency` is straightforward, I know the module name is the same but I don't think it is problematic: `from mne.time_frequency import time_frequency`

kingjr commented:
@mmagnuski 
* ok I changed `morlet` into `_make_morlet` and `_dpss_wavelet` into `_make_dpss`
* I agree, I changed it for `timefreq_transform`. IMO we shouldn't call a function with the module name. Also, I'd like to keep the `_transform` suffix, because I'd like to make it the convention of the transformers associated with the decoding module. @agramfort WDYT?

mmagnuski commented:
`timefreq_transform` sounds good to me

kingjr commented:
Still same docstring error. @Eric89GXL do you know what I should do?

Eric89GXL commented on mne/time_frequency/tfr.py:
```diff
@@ -344,36 +616,17 @@ def cwt(X, Ws, use_fft=True, mode='same', decim=1):
-
 @verbose
+@deprecated("This function will be removed in mne 0.14; use mne.time_frequency"
+            ".timefreq_transform() instead.")
```
Put deprecated before verbose and it should fix the docstring check, we introspect for deprecated decorator but not dep under verbose

Eric89GXL commented:
FYI `flake8` has some new bug where `# noqa` is no longer respected at the end of multi-line strings, I opened an issue for them:

https://gitlab.com/pycqa/flake8/issues/177

kingjr commented:
@Eric89GXL ok, I don't manage to reproduce the flake8 error (which isn't very explicit?). I don't think there's any other issue, so I'm putting it to merge

Eric89GXL commented:
It says the lines are too long, which they are. But we have a `# noqa` at the end of the line that should be respected but isn't. Make sense?

You'd need to do `pip install --upgrade flake8` to see the error locally.

kingjr commented:
@Eric89GXL yes, I've upgraded flake8 (but with conda), can't reproduce the error. I have manually checked that a # noqa was adequately added for the lines > 80. Is there anything else I should do?

Eric89GXL commented:
You have to do it with `pip` to see the error, but don't worry about it / those. It has been fixed already by the Flake8 folks, so hopefully they'll update PyPi soon.

kingjr commented:
Thanks @Eric89GXL 
@agramfort LMKWYT

choldgraf commented:
I don't have time to take a look at this PR right now but I just stumbled upon it and I think it's a great idea to refactor this module :)

A quick question: is the hope to eventually support a high-level class for representing a "full" TFR decomposition? (data that is `shape (epochs, channels, frequencies, times)`, obv not for this PR)?

kingjr commented:
@choldgraf I think it'd make sense indeed, to keep the freqs, time chans etc attached to the data.

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -43,6 +43,8 @@ Changelog
 
     - Add the ``--no-decimate`` option to :ref:`mne make_scalp_surfaces <gen_mne_make_scalp_surfaces>` to skip the high-resolution surface decimation step, by `Eric Larson`_
 
+    - Added :func:`mne.time_frequency.timefreq_transform` and deprecated :func:`mne.time_frequency.morlet`, :func:`mne.time_frequency.cwt_morlet` and :func:`mne.time_frequency.single_trial_power`, to compute time frequency transforms on arrays, by `Jean-Remi King`_
```
Deprecations should be listed in the API section

Eric89GXL commented on mne/time_frequency/tfr.py:
```diff
@@ -29,52 +30,73 @@
+# Make wavelet
+
+@deprecated("This function will be removed in mne 0.14.")
+def morlet(sfreq, freqs, n_cycles=7.0, sigma=None, zero_mean=False):
```
Why are we deprecating this? I don't think we had support issues with it, and [it is used in an example] (https://github.com/mne-tools/mne-python/blob/master/tutorials/plot_background_filtering.py#L103).

kingjr replied:
dpss was private but morlet wasn't, so for consistency. In the previous commit, I made both of them public, but I think @agramfort wanted to diminish the number of low level function related to time frequency transform. 

I have indeed missed this tutorials. I can put it back to public, although it seems a pretty low level function for MNE. It looks like more something relevant to scipy.signal no?

Eric89GXL replied:
> It looks like more something relevant to scipy.signal no?

Probably, but the development timeline there is longer, and I doubt we have anyone willing to do the PR. I'd say leave the existing one alone -- even though it's inconsistent with DPSS, it has been that way a long time without creating problems AFAIK

kingjr replied:
Ok, point taken

Eric89GXL commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +283,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    n_cycles : float | array of float, defaults to 7.0
+        Number of cycles  in the Morlet wavelet. Fixed number
+        or one per frequency.
+    zero_mean : bool
```
| None

Eric89GXL replied:
... defaults to None, which will use Tre for method='mtm' and False for method='morlet'.

Eric89GXL commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +283,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        decomposition.
+        If `int`, returns tfr[..., ::decim].
+        If `slice` returns tfr[..., decim].
+        Note that decimation may create aliasing artifacts.
```
`.. note:`

Eric89GXL commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +283,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        If 'phase', single trial phase.
+        If 'avg_power', average of single trial power.
+        If 'itc', inter-trial coherence.
+        If 'avg_power_itc', average of single trial power and inter-trial coherence across trials. # noqa
```
I think this will render as a single line, can you check CircleCI?

I'd go with a bulleted list probably.

kingjr commented:
@cmoutard @jona-sassenhagen @choldgraf Could you guys give it a try to `timefreq_transform()` to check that your memory/CPU don't blow up in your typical use cases?

```
import numpy as np
from mne.time_frequency import timefreq_transform

n_trials, n_chans, n_times, sfreq = 200, 200, 1000, 256.
frequencies = np.logspace(1, 2, 10)
X = np.random.randn(n_trials, n_chans, n_times)
tfr = timefreq_transform(X, frequencies, sfreq, method='morlet',
                         output='power', decim=20, n_jobs=-1)
```

dengemann commented:
@kingjr `timefreq_transform`sounds matlabish ... can we come up with another name? `time_frequency_tranform` `tfr_transform` etc.

kingjr commented:
Suggestions welcome. tfr_transform has already been ruled out. time_frequency_transform seems a bit long.

dengemann commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +248,244 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    sfreq : float | int, defaults to 1.0
+        Sampling frequency of the data.
+    method : 'mtm' | 'morlet', defaults to 'morlet'
+        The time frequency method. 'morlet' convolves a Morlet wavelet. 'mtm'
```
we already call 'mtm' 'multitaper' elsewhere, we should be consistent here if possible.

dengemann commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +248,244 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        raise ValueError('frequencies must be of shape n_freqs.')
+
+    # Check sfreq
+    if not isinstance(sfreq, (float, int)):
```
this can crash in python 3.5 for numpy types, to be kept in mind. Another test could be `np.isreal`. Probably it would be unlikely not to pass pure python types here but in other places it actually can pose problems.

kingjr replied:
np.isreal doesn't raise None,.

It works like this, so I'm keeping it. Let's not fix what's not broken ;)

dengemann commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +248,244 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+        raise ValueError('method must be "morlet" or "mtm"')
+
+    # Default zero_mean = True if mtm else False
+    zero_mean = method == 'mtm' if zero_mean is None else zero_mean
```
`zero_mean = (method == 'mtm' if zero_mean is None else zero_mean)` would be a more readable local convention

dengemann commented on mne/time_frequency/tfr.py:
```diff
@@ -245,6 +248,244 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    # Check wavelets
+    if len(Ws[0][0]) > epoch_data.shape[2]:
+        raise ValueError('At least one of the wavelets is longer than the '
+                         'signal. Use a longer signal or shorter wavelets.')
```
didn't we have a low level check for this inside `morlet` ? `#consistency`

kingjr replied:
`morlet()` creates the wavelet base without knowing what it will be convolved with. The check has to be done here IMO.

dengemann replied:
Do you remember the message that you get when computing average tfrs with morlet? it must be a function that calls `morlet`

kingjr replied:
I'm not sure to understand the question but the previous error message was 'Wavelet is too long for such a short signal. '

dengemann commented:
> Suggestions welcome. tfr_transform has already been ruled out. time_frequency_transform seems a bit long.

Another avoidable problem with the proposed name is that it breaks with the Python convention of using verbs in functions while using nouns in classes. Some more suggestions:

- `compute_time_frequency`
- `compute_tfr`
- `transform_tfr`
- `map_tfr`
- `map_time_frequency`
- `get_*`
- `filter_*`

jona-sassenhagen commented:
I like the idea of having transform in the name, although it might be overpromising.

agramfort commented:
@kingjr can you rename to compute_tfr (consistent with other function naming eg compute_source_csd and compute_* functions in http://martinos.org/mne/dev/python_reference.html)

also you need to rebase.

thanks heaps

mmagnuski commented:
I like `compute_tfr` :) (but prefer `compute_time_frequency` :P))

agramfort commented:
we use tfr in tfr_morlet etc. so +1 for compute_tfr

kingjr commented:
@agramfort Following our discussion, shall I make it public, merge and you open a PR to decide whether you make it private etc?

agramfort commented:
I am taking a look now

kingjr commented:
@agramfort travis & cis are happy.

agramfort commented:
@kingjr here is a gist to replicate the bug

```
import numpy as np
from mne.time_frequency import _compute_tfr

sfreq = 200.0
n_times = int(sfreq)  # Second long epochs
n_epochs = 3
seed = 42
rng = np.random.RandomState(seed)
noise = 0.1 * rng.randn(n_epochs, 4, n_times)
t = np.arange(n_times, dtype=np.float) / sfreq
signal = np.sin(np.pi * 2. * 50. * t)  # 50 Hz sinusoid signal
signal[np.logical_or(t < 0.45, t > 0.55)] = 0.  # Hard windowing
on_time = np.logical_and(t >= 0.45, t <= 0.55)
signal[on_time] *= np.hanning(on_time.sum())  # Ramping
data = noise + signal

freqs = np.arange(5, 100, 3, dtype=np.float)
n_cycles = freqs / 2.
n_jobs = 1
use_fft = False
decim = 2
time_bandwidth = 4.

out_avg = _compute_tfr(data, freqs, sfreq, n_cycles=n_cycles,
                       n_jobs=n_jobs, use_fft=use_fft, decim=decim,
                       zero_mean=True, method='mtm',
                       time_bandwidth=time_bandwidth, output='avg_power')

out = _compute_tfr(data, freqs, sfreq, n_cycles=n_cycles,
                   n_jobs=n_jobs, use_fft=use_fft, decim=decim,
                   zero_mean=True, method='mtm',
                   time_bandwidth=time_bandwidth, output='complex')

print(out_avg - np.mean(np.abs(out) ** 2, axis=0))
```

kingjr commented:
Good catch. So correcting _time_frequency_loop by
```
            # Stack or add
            if ('avg_' in output) or ('itc' in output):
                tfrs += tfr
            else:
                tfrs[epoch_idx] += tfr  # HERE WE ADD ACROSS TAPERS
```

makes the mean(single trial power)  == avg_power 

```

out_avg = compute_tfr(data, freqs, sfreq, n_cycles=n_cycles,
                      n_jobs=n_jobs, use_fft=use_fft, decim=decim,
                      zero_mean=True, method='multitaper',
                      time_bandwidth=time_bandwidth, output='avg_power')

out = compute_tfr(data, freqs, sfreq, n_cycles=n_cycles,
                  n_jobs=n_jobs, use_fft=use_fft, decim=decim,
                  zero_mean=True, method='multitaper',
                  time_bandwidth=time_bandwidth, output='power')
print(out_avg - np.mean(out, axis=0))
```

However, for the complex values, *I think* that `mean(abs(complex) ** ` must be different from `avg_power` , because the single trial power is estimated per taper, whereas the complex values are the aggregation of multiple tapers.

We could thus add a NotImplemented error for single trial complex estimate in the same way that I did for the phase.

agramfort commented:
aggregation of complex values across tapers is indeed wrong.

however I need single trial mtm ? can I do this with compute_tfr?

kingjr commented:
you can have single trial multitaper 'power' with compute tfr, and this works, try test above.

kingjr commented:
An alternative would be to estimate the single trial phase and single trial power per taper and recreate the complex values from the aggregate; but this could be an subsequent PR.

agramfort commented:
closing in favor of #3476
